doc/specs/#2046 - Command Palette.md

```diff
@@ -0,0 +1,284 @@
+  connection, followed by a newline (as to execute the command).
+  `suppressNewline` would prevent the newline from being added. This would work
+  relatively well, so long as you're sitting at a shell prompt. If you were in
+  an application like `vim`, this might be handy for executing a sequence of
+  vim-specific keybindings. Otherwise, you're just going to end up writing a
+  commadline to the buffer of vim. It would be weird, but not unexpected.
+* Additionally mentioned in [#2046] was the potential for profile-scoped
+  commands. While that's a great idea, I believe it's out of scope for this
+  spec.
+* Once [#754] lands, we'll need to make sure to include commands for each action
+  manually in the default settings. This will add some overhead that the
+  developer will need to do whenever they add an action. That's unfortunate, but
+  will be largely beneficial to the end user.
+* We could theoretically also display the keybinding for a certain command in
+  the `ListViewItem` for the command. We'd need some way to correlate a
+  command's action to a keybinding's action. This could be done in a follow-up
+  task.
+* We might want to alter the fuzzy-search algorithim, to give higher precedence
+  in the results list to commands with more consecutive matching characters.
+  Alternatively we could give more weight to commands where the search matched
+  the initial character of words in the command.
+  - For example: `ot` would give more weight to "**O**pen **T**ab" than
+    "**O**pen Se**t**tings").
+* Another idea for a future spec is the concept of "nested commands", where a
+  single command has many sub-commands. This would hide the children commands
+  from the entire list of commands, allowing for much more succinct top-level
+  list of commands, and allowing related commands to be grouped together.
+  - For example, I have a text editor plugin that enables rendering markdown to
+    a number of different styles. To use that command in my text editor, first I
+    hit enter on the "Render Markdown..." command, then I select which style I
+    want to render to, in another list of options. This way, I don't need to
+    have three options for "Render Markdown to github", "Render Markdown to
+    gitlab", all in the top-level list.
+  - We probably also want to allow a nested command set to be evaluated at
+    runtime somehow. Like if we had a "Open New Tab..." command that then had a
+    nested menu with the list of profiles.
+* We may want to add a button to the New Tab Button's dropdown to "Show Command
+  Palette". I'm hesitant to keep adding new buttons to that UI, but the command
+  palette is otherwise not highly discoverable.
+  - We could add another button to the UI to toggle the visibility of the
+    command palette. This was the idea initially proposed in [#2046].
+  - For both these options, we may want a global setting to hide that button, to

```

Comment:
What if a user wants to add a command to their _menu_? There's been requests for custom menus.